Refactor: libpe_status: Bring common_unpack() up to current guidelines

nrwahl2 opened the pull request:
* Rename to `pe__unpack_resource()` and make library-private
* Use `true`/`false` instead of `TRUE`/`FALSE` with its `bool` variables
* Return a standard Pacemaker return code (currently only `pcmk_rc_ok` or `pcmk_rc_error`, since no others seemed to fit)
* Add a Doxygen comment block

Closes T540

kgaillot commented on lib/pengine/complex.c:
```diff
@@ -596,16 +596,16 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
     if (id == NULL) {
         pe_err("Must specify id tag in <resource>");
-        return FALSE;
+        return pcmk_rc_error;
```
We should avoid pcmk_rc_error except in dire circumstances (generally when we don't know what happened). For these, pcmk_rc_schema_validation or pcmk_rc_cib_corrupt would be acceptable. Since the callers only check vs pcmk_rc_ok it's not terribly important, but we want to give something reasonable if someone ever calls pcmk_rc_str() on the result.

In general, if we know the cause of the error, adding a new pcmk_rc_ code is preferable to using pcmk_rc_error.

nrwahl2 replied:
Updated

kgaillot commented on lib/pengine/complex.c:
```diff
@@ -570,6 +570,22 @@ unpack_requires(pe_resource_t *rsc, const char *value, bool is_default)
+ * \param[in]  xml_obj   XML node containing the resource's configuration
+ * \param[out] rsc       Destination for unpacking
+ * \param[...] parent    Resource's parent (\c NULL if the resource is a
+ *                       primitive)
```
in is sufficient here. The pointer can't be const since we're assigning it to a non-const struct member, but this function can't change it.

nrwahl2 replied:
That's just a typo -- I made placeholders and forgot to fill that one in.

kgaillot commented on lib/pengine/complex.c:
```diff
@@ -570,6 +570,22 @@ unpack_requires(pe_resource_t *rsc, const char *value, bool is_default)
+ * \param[out] rsc       Destination for unpacking
+ * \param[...] parent    Resource's parent (\c NULL if the resource is a
+ *                       primitive)
+ * \param[in]  data_set  Cluster working set
```
in,out following the newly established convention, since this function can change data_set members via called functions

nrwahl2 replied:
Ah, gotcha. I wasn't sure where we landed on that.

kgaillot reviewed (commented):
looking good

nrwahl2 commented:
General question: Current preference on the use or non-use of `ID()` and `TYPE()` macros in `msg_xml.h`? I get the impression they've fallen out of favor despite not being deprecated yet.

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -596,16 +596,15 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
     if (id == NULL) {
         pe_err("Must specify id tag in <resource>");
-        return FALSE;
+        return pcmk_rc_schema_validation;
```
The common existing use cases of `pcmk_rc_schema_validation` throughout Pacemaker are much like this one: return this error if something about the XML is invalid or doesn't align with our expectations, like a missing attribute or an attribute that refers to something it shouldn't.

With that said, it doesn't align with the documented use case of `pcmk_rc_schema_validation`: `"Update does not conform to the configured schema"`. Eventually, it would be nice to either clean these up or make the message more inclusive. For now, I'm using this RC the way it's often used elsewhere.

kgaillot replied:
I think your new code is better throughout

nrwahl2 replied:
I'm taking this to mean use `pcmk_rc_unpack_error` for all five error-returns.

kgaillot replied:
yep

kgaillot replied:
and for a lot of the preexisting uses

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -632,7 +631,7 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
         pe_err("Unknown resource type: %s", crm_element_name((*rsc)->xml));
         free(*rsc);
-        return FALSE;
+        return pcmk_rc_schema_validation;
```
See above, under the case of `"Must specify id tag in <resource>"`. This isn't really a schema validation error but it fits the existing pattern.

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -596,16 +596,15 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
         pe_err("Nowhere to unpack resource into");
-        return FALSE;
-
+        return EINVAL;
```
A null destination pointer is an invalid argument.

kgaillot replied:
It would be worthwhile to do this check first, using CRM_CHECK() rather than pe_err(). pe_err() will log a configuration error whereas this would be a developer mistake.

nrwahl2 replied:
Done. `CRM_CHECK()` is used quite a bit for this situation. I would have expected `CRM_ASSERT()` if the intent is to catch a developer error.

kgaillot replied:
True, assert would be fine in this case

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -596,16 +596,15 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
 
     if (unpack_template(xml_obj, &expanded_xml, data_set) == FALSE) {
-        return FALSE;
+        return pcmk_rc_unpack_error;
```
`unpack_template()` can fail due to a schema error or due to certain XML elements not found
* Missing object ID
* Improper reference to self
* No resources configured
* Named template does not exist in XML

A case could be made that nonexistent elements are still `pcmk_rc_schema_validation` errors, since other elements refer to or imply the existence of these nonexistent elements.

When `unpack_template()` is updated to return a standard Pacemaker return code later, we can just return its RC.

kgaillot replied:
Yep

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -848,7 +847,7 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
 
     if ((*rsc)->fns->unpack(*rsc, data_set) == FALSE) {
-        return FALSE;
+        return pcmk_rc_unpack_error;
```
The causes of failure vary, so we can't put them into one neat box.
```
Native:
* Resource of given standard can't be used as promotable clone

Group:
* Always returns TRUE

Clone:
* Empty child object
* Failure to create child

Bundle:
* No container child
* Invalid network
* Call to pe__unpack_resource() fails
* Container creation failure
```

kgaillot replied:
Yep, once we can change unpack() to return a standard code we can propagate that here. Unfortunately that's public API so it will have to wait a while.

nrwahl2 replied:
:'(

nrwahl2 commented on lib/pengine/complex.c:
```diff
@@ -873,10 +872,10 @@ pe__unpack_resource(xmlNode *xml_obj, pe_resource_t **rsc,
     if (expanded_xml) {
         if (add_template_rsc(xml_obj, data_set) == FALSE) {
-            return FALSE;
+            return pcmk_rc_unpack_error;
```
`add_template_rsc()`'s direct failure conditions would be `pcmk_rc_schema_validation`, but it calls `add_tag_ref()`, which returns `FALSE` for memory allocation failure.

kgaillot replied:
Yep, ditto for add_template_rsc()'s return value. When we know something is a memory error, ENOMEM is appropriate, but when a memory error is mixed in with other possibilities, it's fine to ignore the memory case and focus on the rest, because a memory issue will very quickly result in a fatal exit and/or fencing, and it might not even be possible to log a message.

nrwahl2 replied:
Ah, fair enough.

nrwahl2 commented:
For `pcmk_rc2exitc()` for `pcmk_rc_unpack_error`, I let it default to `CRM_EX_ERROR`. We could do a new exit code if we want.

Related: `EINVAL` is mapped to `CRM_EX_SOFTWARE`. Seems like maybe a better fit for `CRM_EX_INVALID_PARAM`.

kgaillot commented:
> General question: Current preference on the use or non-use of `ID()` and `TYPE()` macros in `msg_xml.h`? I get the impression they've fallen out of favor despite not being deprecated yet.

They're fine. Eventually I'd like to rename them per current guidelines but the functionality is useful.

kgaillot commented:
> For `pcmk_rc2exitc()` for `pcmk_rc_unpack_error`, I let it default to `CRM_EX_ERROR`. We could do a new exit code if we want.

I'm thinking CRM_EX_CONFIG makes sense. CRM_EX_NOT_CONFIGURED and CRM_EX_DATAERR are also in the ballpark.


> 
> Related: `EINVAL` is mapped to `CRM_EX_SOFTWARE`. Seems like maybe a better fit for `CRM_EX_INVALID_PARAM`.

We typically use EINVAL for developer errors (misused functions), rather than invalid user-supplied values.

kgaillot commented on lib/pengine/complex.c:
```diff
@@ -570,6 +570,21 @@ unpack_requires(pe_resource_t *rsc, const char *value, bool is_default)
+ * \param[in]     xml_obj   XML node containing the resource's configuration
+ * \param[out]    rsc       Where to store the unpacked resource information
+ * \param[in]     parent    Resource's parent (\c NULL if the resource is a
+ *                          primitive)
```
parent is NULL for top-level resources of any type. E.g. for a top-level group, the group's parent is NULL, and the individual primitive resources inside it have the group as the parent. For a cloned group, the clone's parent is NULL, the group's parent is the clone, and the group members' parent is the group.

nrwahl2 replied:
Oh yeah, duh. Thanks.

kgaillot reviewed (commented):
looking good

nrwahl2 commented:
Addressed review:
* Use `CRM_CHECK` -> `EINVAL` for null `rsc` pointer.
* Return `pcmk_rc_unpack_error` for other failure conditions
* Updated description to "Unable to parse CIB XML" and mapped to `CRM_EX_CONFIG`.
* Fixed doc comment about primitive `rsc` null pointer.
* Replace many existing occurrences of `pcmk_rc_schema_validation` with `EINVAL` or `pcmk_rc_unpack_error`.
